From: Ursula Braun <ursula.braun@de.ibm.com>
Subject: af_iucv: avoid path quiesce of severed path in shutdown()
Patch-mainline: v3.17-rc1
Git-commit: 1042cab8627a2d11491e8b0dd40c4dda3180285a
References: bnc#954986, LTC#131684

Description:  af_iucv: avoid path quiesce of severed path in shutdown()
Symptom:      sendmsg() returns with -EPIPE during a stress test
Problem:      This is caused by quiescing a path even though it has
              been already severed by peer. For IUCV transport shutdown()
              consists of 2 steps:
              (1) sending the shutdown message to peer
              (2) quiescing the iucv path
              If the iucv path between these 2 steps is severed due to peer
              closing the path, the quiesce step is no longer needed.
Solution:     In iucv_sock_shutdown() make sure path_quiesce() is called only
              if the iucv path is still available.
Reproduction: Run af_iucv stress test for af_iucv sockets with IUCV transport.

Upstream-Description:

              af_iucv: avoid path quiesce of severed path in shutdown()

              An af_iucv stress test showed -EPIPE results for sendmsg()
              calls. They are caused by quiescing a path even though it has
              been already severed by peer. For IUCV transport shutdown()
              consists of 2 steps:
              (1) sending the shutdown message to peer
              (2) quiescing the iucv path
              If the iucv path between these 2 steps is severed due to peer
              closing the path, the quiesce step is no longer needed.

              Signed-off-by: Ursula Braun <ursula.braun@de.ibm.com>
              Signed-off-by: Frank Blaschka <blaschka@linux.vnet.ibm.com>
              Reported-by: Philipp Hachtmann <phacht@linux.vnet.ibm.com>
              Signed-off-by: David S. Miller <davem@davemloft.net>


Signed-off-by: Ursula Braun <ursula.braun@de.ibm.com>
Acked-by: John Jolly <jjolly@suse.de>
---
 net/iucv/af_iucv.c |    3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

--- a/net/iucv/af_iucv.c
+++ b/net/iucv/af_iucv.c
@@ -1534,7 +1534,8 @@ static int iucv_sock_shutdown(struct soc
 
 	sk->sk_shutdown |= how;
 	if (how == RCV_SHUTDOWN || how == SHUTDOWN_MASK) {
-		if (iucv->transport == AF_IUCV_TRANS_IUCV) {
+		if ((iucv->transport == AF_IUCV_TRANS_IUCV) &&
+		    iucv->path) {
 			err = pr_iucv->path_quiesce(iucv->path, NULL);
 			if (err)
 				err = -ENOTCONN;
